### PR TITLE
Unit tests backend

### PR DESCRIPTION
**Pull request by jdtoombs**
Splitting off some of the tests instead of submitting one big pile of unit tests - some of these tests are still being worked at to achieve 100% coverage for the service

### CODE REVIEW

File: backend/tests/unit/dal/Services/Admin/AddressServiceTest.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+            var updateAddress= EntityHelper.CreateAddress(1, "1234 Street", null, "V9C9C9");;
+            var newName = "1234 Updated Street";
+            updateAddress.Address1 = newName;
+
+            // Act
+            service.Update(updateAddress);
+
+            // Assert
+            address.Address1.Should().Be(newName);
```
Comment (Fosol) [inline backend/tests/unit/dal/Services/Admin/AddressServiceTest.cs:47]: I wonder if you could find a way to verify the following too?

- The state of the entity address was changed to `EntityState.Modified`
- The `Context.CommitTransaction()` was called

File: backend/tests/unit/dal/Services/Admin/AgencyServiceTest.cs [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+            using var init = helper.InitializeDatabase(user);
+
+            var service = helper.CreateService<AgencyService>(user);
+            var context = helper.GetService<PimsContext>();
+
+            // Act
+            var result = service.GetAll();
+
+            // Assert
+            Assert.Equal(18, result.Count());
```
Comment (Fosol) [inline backend/tests/unit/dal/Services/Admin/AgencyServiceTest.cs:124]: Might be wise to generate the count dynamically so this doesn't fail when someone updates the InitializeDatabase

File: backend/tests/unit/dal/Services/Admin/ProjectServiceTest.cs [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+
+            // Act
+            var result = service.GetSnapshots(1);
+
+            // Assert
+            Assert.NotNull(result);
+            Assert.Equal(expectedCount, result.Count());
+        }
+
+                [Fact]
```
Comment (Fosol) [inline backend/tests/unit/dal/Services/Admin/ProjectServiceTest.cs:116]: spacing?

File: backend/tests/unit/dal/Services/Admin/PropertyClassificationTest.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+            var service = helper.CreateService<PropertyClassificationService>(user);
+            var context = helper.GetService<PimsContext>();
+
+            // Act
+            var result = service.GetAll();
+
+            // Assert
+            // There are 8 classificaitons by default
+            Assert.True(result.Count() > 0);
```
Comment (Fosol) [inline backend/tests/unit/dal/Services/Admin/PropertyClassificationTest.cs:45]: Not a great test lol

File: backend/tests/unit/dal/Services/Admin/ProvinceServiceTest.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            var user = PrincipalHelper.CreateForPermission(Permissions.SystemAdmin);
+            using var init = helper.InitializeDatabase(user);
+
+            var service = helper.CreateService<ProvinceService>(user);
+
+            // Act
+            var result = service.Get();
+
+            // Assert
+            Assert.True(result.Count() > 0);
```
Comment (Fosol) [inline backend/tests/unit/dal/Services/Admin/ProvinceServiceTest.cs:42]: Not a great test.  You can dynamically generate the expected count.

File: backend/tests/unit/dal/Services/Admin/ProvinceServiceTest.cs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            helper.CreatePimsContext(user, true).AddAndSaveChanges(province);
+
+            var service = helper.CreateService<ProvinceService>(user);
+            var context = helper.GetService<PimsContext>();
+
+            // Act
+            service.Remove(province);
+
+            // Assert
+            Assert.Equal(EntityState.Detached, context.Entry(province).State);
```
Comment (Fosol) [inline backend/tests/unit/dal/Services/Admin/ProvinceServiceTest.cs:120]: This kind of assert could be applied to other add/update tests to make them more robust.

**Review by Fosol — APPROVED** [review]
Looks great!  Just a few comments to improve.

File: backend/tests/unit/dal/Services/Admin/ProjectServiceTest.cs (hunk 3 above)
Comment (jdtoombs) [reply backend/tests/unit/dal/Services/Admin/ProjectServiceTest.cs:116]: Yeah trying to get my formatter to behave lol, I'll fix before I merge

File: backend/tests/unit/dal/Services/Admin/ProvinceServiceTest.cs (hunk 6 above)
Comment (jdtoombs) [reply backend/tests/unit/dal/Services/Admin/ProvinceServiceTest.cs:120]: Gotchya, thanks! Will adjust the other tests. Will also fix the `result.Count()` asserts.

File: backend/tests/unit/dal/Services/Admin/ProjectServiceTest.cs [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+            // Assert
+            Assert.Throws<NotAuthorizedException>(() =>
+                service.Add(project));
+        }
+        #endregion
+
+
+        #region Update
+        [Fact]
+        public void Update()
```
Comment (devinleighsmith) [inline backend/tests/unit/dal/Services/Admin/ProjectServiceTest.cs:211]: no keynotfound, unauthorized for update?

File: backend/tests/unit/dal/Services/Admin/PropertyClassificationTest.cs (hunk 4 above)
Comment (devinleighsmith) [reply backend/tests/unit/dal/Services/Admin/PropertyClassificationTest.cs:45]: can probably compare against the number of classifications in the context.

File: backend/tests/unit/dal/Services/Admin/PropertyClassificationTest.cs [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            var helper = new TestHelper();
+            var user = PrincipalHelper.CreateForPermission();
+            var service = helper.CreateService<PropertyClassificationService>(user);
+            var classification = EntityHelper.CreatePropertyClassification("TEST");
+            // Act
+            // Assert
+            Assert.Throws<KeyNotFoundException>(() =>
+                service.Update(classification));
+        }
+        #endregion
```
Comment (devinleighsmith) [inline backend/tests/unit/dal/Services/Admin/PropertyClassificationTest.cs:76]: no success test case for update?

File: backend/tests/unit/dal/Services/Admin/ProvinceServiceTest.cs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+            // Act
+            var result = service.Get();
+
+            // Assert
+            Assert.True(result.Count() > 0);
+        }
+        #endregion
+        #region Update
+        [Fact]
+        public void Update_Province()
```
Comment (devinleighsmith) [inline backend/tests/unit/dal/Services/Admin/ProvinceServiceTest.cs:47]: do we need auth tests for this service?

File: backend/tests/unit/dal/Services/Admin/ProvinceServiceTest.cs [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+            // Act
+            service.Remove(province);
+
+            // Assert
+            Assert.Equal(EntityState.Detached, context.Entry(province).State);
+        }
+
+        [Fact]
+        public void Remove_Province_KeyNotFound()
```
Comment (devinleighsmith) [inline backend/tests/unit/dal/Services/Admin/ProvinceServiceTest.cs:124]: do we need a null argument test?

File: backend/tests/unit/dal/Services/Admin/RoleServiceTest.cs [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+            // Assert
+            Assert.NotNull(result);
+            Assert.IsAssignableFrom<Pims.Dal.Entities.Role>(result);
+            Assert.Equal("Role 1", result.Name);
+        }
+        #endregion
+        #region Update
+        [Fact]
+        public void Update_Role()
```
Comment (devinleighsmith) [inline backend/tests/unit/dal/Services/Admin/RoleServiceTest.cs:148]: do we need error case tests for update/remove?

File: backend/tests/unit/dal/Services/Admin/UserServiceTest.cs [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+            var result = service.Get(id);
+
+            // Assert
+            Assert.NotNull(result);
+            Assert.Equal(id, result.Id);
+        }
+        #endregion
+        #region Add
+        [Fact]
+        public void Add_User()
```
Comment (devinleighsmith) [inline backend/tests/unit/dal/Services/Admin/UserServiceTest.cs:141]: unauth tests for this service?

**Review by devinleighsmith — APPROVED** [review]

File: backend/tests/unit/dal/Services/Admin/UserServiceTest.cs (hunk 12 above)
Comment (jdtoombs) [reply backend/tests/unit/dal/Services/Admin/UserServiceTest.cs:141]: Yeah, a number of these tests are still in progress, was just throwing up some of the ones I had to make sure I wasn't completely off the mark. I am tracking them all with codecov!

**Pull request merged**